Components: refactor Button input handling

Be-ing opened the pull request:
Remove the old onlyOnPress Button property. When set to false, this would toggle the value of a Control on button up and button down, which is [not really the correct behavior](https://github.com/mixxxdj/mixxx/pull/1184#discussion_r100699865). Instead, let Buttons have a type, with the default type being push. The button types are:

  * push: set inKey to 1 on button press and 0 on button release
  * toggle: invert value of inKey on button press
  * powerWindow: like toggle, but toggles the value of inKey again on button up when long pressed

The ComponentContainer.shift() and ComponentContainer.unshift() functions automatically take care of setting the inKey of push type Buttons to 0 on shift/unshift to ensure that it does not matter whether shift or the Button is released first. Before this [required some ugly hacks](https://github.com/mixxxdj/mixxx/pull/1243#issuecomment-304557891) that would have to be implemented by each Button.

This is a backwards incompatible change. Previously, the default was to set the onlyOnPress property to `true`, equivalent to the new toggle type. Buttons that had onlyOnPress set to `false` before can simply remove that and use the default push type. All the special Buttons implemented in the library such as HotcueButton and the Buttons for the EffectUnit have been updated appropriately. I'll update #1200 and #1243 for the new API.

Be-ing committed:
Components: refactor Button input handling

remove the old onlyOnPress Button property. When set to false, this
would toggle the value of a Control on button up and button down, which
is not really the correct behavior. Instead, let Buttons have a type,
with the default type being push. The button types are:

push: set inKey to 1 on button press and 0 on button release
toggle: toggle value of inKey on button press
trigger: set inKey to 1 on button press
powerWindow: like toggle, but toggles the value of inKey again on button
up when long pressed

The ComponentContainer.shift() and ComponentContainer.unshift()
functions automatically take care of setting the inKey of push type
Buttons to 0 on shift/unshift to ensure that it does not matter whether
shift or the Button is released first.

Be-ing committed:
Components: don't reset inKey to 0 on shfit for specialized Buttons

that have specialized input functions instead of the prototype
Button input function

Be-ing commented:
@ronso0 and @uklotzde, any thoughts on this?

ronso0 commented:
I can test this next days.
Essentially, now the physical FX enable buttons would work like respective buttons in skin, right?

Be-ing commented:
>  Essentially, now the physical FX enable buttons would work like respective buttons in skin, right?

That has already been the case since #1249 was merged. The EffectUnit API has not changed. I'm looking for feedback on the changes to the Button API.

Be-ing committed:
Components: remove trigger Button type

I do not think there is a use case for it. Controls that only do
something when set to 1 still need to be reset to 0 or their button in
the skin will remain on after releasing the controller Button.

Be-ing committed:
Components: powerWindow Button type code cleanup

remove log spam from killing expired timers

Be-ing commented:
Can we think of a better name than powerWindow?

ronso0 commented:
> That has already been the case since #1249 was merged. The EffectUnit API has not changed, neither has its behavior.

Woohoo, wasn't aware of that PR. Thanks!
I use Components for effects only.

daschuer commented:
@ronso0 can we merge this now?

ronso0 commented:
@daschuer Sorry, I can't tell...as I said I use Components for effects only, don't know what difference it makes for a complete Components mapping

daschuer commented:
Ok, than I think we can merge this, and fix pending issues in master if
there are any. Right?


Am 31.05.2017 3:11 vorm. schrieb "ronso0" <notifications@github.com>:

@daschuer <https://github.com/daschuer> Sorry, I can't tell...as I said I
use Components for effects only, don't know what difference it makes for a
complete Components mapping

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/mixxxdj/mixxx/pull/1271#issuecomment-305053288>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABsfInp35O4xJyaxlDwk2obpJCQa7kzaks5r_L47gaJpZM4No0e9>
.

Be-ing commented:
Sure, then #1200 and #1243 can be merged, but those need #940 merged to see beatloop_size and beatjump_size on screen.

daschuer commented:
Ok, Thank you for your work.
